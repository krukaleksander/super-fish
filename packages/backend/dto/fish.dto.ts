import {IsBoolean, IsNotEmpty, IsNumber, IsOptional, IsString, ValidateNested} from "class-validator";
import {Type} from "class-transformer";
import {IFish, IFishPackage} from "../src/interfaces";

export class FishDto implements IFish {
  @IsOptional()
  @IsNumber()
  id?: string;
  @IsString()
  @IsNotEmpty()
  frontSide: string;
  @IsString()
  @IsNotEmpty()
  backSide: string;
  @IsBoolean()
  @IsOptional()
  remember?: boolean;
  @IsString()
  @IsOptional()
  createdAt?: string;
  @IsString()
  @IsOptional()
  refreshedAt?: string;
}

export class FishPackageDto implements IFishPackage {
  @IsOptional()
  @IsNumber()
  id?: string;
  @IsString()
  @IsNotEmpty()
  name: string;
  @IsOptional()
  @ValidateNested({each: true})
  @Type(() => FishDto)
  shoalOfFish?: FishDto[];
}
