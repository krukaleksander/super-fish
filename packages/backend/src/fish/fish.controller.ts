import {Body, Controller, Delete, Get, HttpCode, Post, Put} from '@nestjs/common';
import {FishService} from "./fish.service";
import {FishDto, idOfDeletingItemDto, nameOfNewPackageDto} from "../../dto/fish.dto";
import {ISendAllFishes} from "../interfaces";

@Controller('fish')
export class FishController {
  constructor(private readonly fishService: FishService) {
  }
  @Get()
  findAll(): ISendAllFishes {
    return this.fishService.findAll();
  }
  @Post()
  createFish(@Body() fish: FishDto) {
    return this.fishService.createFish(fish);
  }
  @Post('/package')
  createPackage(@Body() packageName: nameOfNewPackageDto) {
    return this.fishService.createPackage(packageName);
  }
  @Delete()
  @HttpCode(202)
  deleteFish(@Body() fishId: idOfDeletingItemDto) {
    return this.fishService.deleteFish(fishId)
  }
  @Put()
  updateFish(@Body() fishToUpdate: FishDto) {
    return this.fishService.updateFish(fishToUpdate)
  }
}
