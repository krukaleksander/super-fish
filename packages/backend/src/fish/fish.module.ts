import {Module} from "@nestjs/common";
import {FishController} from "./fish.controller";
import {FishService} from "./fish.service";

@Module({
  controllers: [FishController],
  providers: [FishService]
})

export class FishModule {}
