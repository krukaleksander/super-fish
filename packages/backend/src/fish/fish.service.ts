import { Injectable } from '@nestjs/common';
import {FishDto, FishPackageDto} from '../../dto/fish.dto';
import { IFish, ISendAllFishes } from '../interfaces';
import { v4 as generateID } from 'uuid';
import { mockFishPackage } from '@super-fish/mock-fish-lib';

@Injectable()
export class FishService {
  findAll(): ISendAllFishes {
    return {
      status: 200,
      packages: [mockFishPackage],
    };
  }

  saveFish(fish: FishDto) {
    const { backSide, frontSide, packageID } = fish;
    const newSavedFish: IFish = {
      packageID,
      backSide,
      frontSide,
      id: generateID(),
      remember: false,
      refreshedAt: 'never',
      createdAt: new Date().toString(),
    };
    return {
      status: 201,
      message: 'Fish created',
      fish: newSavedFish,
    };
  }

  createPackage(packageName) {
    const newPackage: FishPackageDto = {
      id: generateID(),
      name: packageName.name,
      shoalOfFish: []
    }
    return {
      status: 201,
      message: 'Package Created',
      package: newPackage
    }
  }
}
