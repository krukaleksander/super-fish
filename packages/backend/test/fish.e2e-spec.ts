import { INestApplication, ValidationPipe } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { FishModule } from '../src/fish/fish.module';
import * as request from 'supertest';
import { FishDto } from '../dto/fish.dto';

describe('Fish', () => {
  let app: INestApplication;

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [FishModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(new ValidationPipe());
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('fish endpoint', () => {
    describe('/fish (GET) all fishes endpoint', () => {
      it("'/fish (GET) response status 200'", () => {
        return request(app.getHttpServer()).get('/fish').expect(200);
      });
      it('/fish (GET) should return all packages of fishes', async () => {
        const { body: response } = await request(app.getHttpServer()).get(
          '/fish'
        );
        expect(response).toMatchSnapshot();
      });
    });
    describe("'/fish (POST) create one fish endpoint'", () => {
      it("'/fish (POST) response status 400 if no fish was send", () => {
        return request(app.getHttpServer()).post('/fish').expect(400);
      });
      it("'/fish (POST) response status 400 if fish is sent but not valid", () => {
        return request(app.getHttpServer())
          .post('/fish')
          .send({
            packageID: '9b1deb4d-3b7d-4bad-9bdd-2b0d7b3dcb6d',
            frontSide: 20,
            backSide: 'Something',
          })
          .expect({
            statusCode: 400,
            message: ['frontSide must be a string'],
            error: 'Bad Request',
          });
      });
      it('/fish (POST) should send back fish details in response', async () => {
        const mockFish: FishDto = {
          backSide: 'Kot',
          frontSide: 'Cat',
          packageID: '1b9d6bcd-bbfd-4b2d-9b5d-ab8dfbbd4bed'
        };
        const response = await request(app.getHttpServer())
          .post('/fish')
          .send(mockFish);
        expect(response.status).toBe(201);
        expect(response.body).toMatchObject({
          status: 201,
          message: 'Fish created',
          fish: {
            frontSide: mockFish.frontSide,
            backSide: mockFish.backSide,
            id: expect.any(String),
            remember: expect.any(Boolean),
            createdAt: expect.any(String),
            refreshedAt: expect.any(String),
          },
        });
      });
    });
    describe('create package endpoint', () => {
      it('should create package and send back details of it', async () => {
        const packageName = '6 minute english 1.2';
        const response = await request(app.getHttpServer())
          .post('/fish/package')
          .send({name: packageName})
        expect(response.status).toBe(201)
        expect(response.body).toMatchObject({
          status: 201,
          message: 'Package Created',
          package: {
            id: expect.any(String),
            name: packageName,
            shoalOfFish: []
          }
        })
      });
      it('should send 400 if package name is not a string', async () => {
        const response = await request(app.getHttpServer())
          .post('/fish/package')
          .send({name: [1,null, undefined]})
        expect(response.status).toBe(400)
      });
      it('should send 400 if package name is too long', async () => {
        const response = await request(app.getHttpServer())
          .post('/fish/package')
          .send({name: 'I\'am writing too long package name =)'})
        expect(response.status).toBe(400)
      });
    })
    describe('delete fish endpoint', () => {
      it('should delete fish if passed id is valid and can be found in db', async () => {
        const response = await request(app.getHttpServer())
          .delete('/fish')
          .send({id: '1b9d6bcd-bbfd-4b2d-9b5d-ab8dfbbd4bed'})
        expect(response.status).toBe(202)
      });
      it('should send 400 if id is not valid', async () => {
        const response = await request(app.getHttpServer())
          .delete('/fish')
          .send({id: true})
        expect(response.status).toBe(400)
      });
    })
  });
});
