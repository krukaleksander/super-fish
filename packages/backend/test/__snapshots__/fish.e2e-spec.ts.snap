// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fish fish endpoint (GET) /fish all fishes endpoint should return all packages of fishes 1`] = `
Object {
  "packages": Array [
    Object {
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b3dcb6d",
      "name": "6 minute english 1.0",
      "shoalOfFish": Array [
        Object {
          "backSide": "difficult to fully understand or comprehend",
          "createdAt": "2022-05-29T06:26:16.174Z",
          "frontSide": "hard to get your head around",
          "id": "1b9d6bcd-bbfd-4b2d-9b5d-ab8dfbbd4bed",
          "packageID": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b3dcb6d",
          "refreshedAt": "never",
          "remember": false,
        },
        Object {
          "backSide": "there are very few of something",
          "createdAt": "2022-05-30T06:26:16.174Z",
          "frontSide": "thin on the ground",
          "id": "b197f765-57cd-4368-8751-cde181e6bfdf",
          "packageID": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b3dcb6d",
          "refreshedAt": "never",
          "remember": false,
        },
      ],
    },
  ],
  "status": 200,
}
`;
